Acceptance Testing Not Compatible After Terraform 0.15.0-alpha20210127

_This is a known issue and is only submitted for discoverability._

### SDK version

```
v2.4.2
```

### Relevant provider source code

Any existing testing code. 😄 

### Expected Behavior

Acceptance testing to continue working as with prior versions of Terraform CLI.

### Actual Behavior

Test failures without workarounds, such as:

```
=== CONT  TestAccAWSAcmCertificateDataSource_multipleIssued
          data_source_aws_acm_certificate_test.go:110: Error running init: Usage: terraform init [options] [DIR]
              
                Initialize a new or existing Terraform working directory by creating
                initial files, loading any remote state, downloading modules, etc.
              
                This is the first command that should be run for any new or existing
                Terraform configuration per machine. This sets up all the local data
                necessary to run Terraform that is typically not committed to version
                control.
              
                This command is always safe to run multiple times. Though subsequent runs
                may give errors, this command will never delete your configuration or
                state. Even so, if you have important information, please back it up prior
                to running this command, just in case.
              
                If no arguments are given, the configuration in this working directory
                is initialized.
              
              Options:
              
                -backend=true        Configure the backend for this configuration.
              
                -backend-config=path This can be either a path to an HCL file with key/value
                                     assignments (same format as terraform.tfvars) or a
                                     'key=value' format. This is merged with what is in the
                                     configuration file. This can be specified multiple
                                     times. The backend type must be in the configuration
                                     itself.
              
                -force-copy          Suppress prompts about copying state data. This is
                                     equivalent to providing a "yes" to all confirmation
                                     prompts.
              
                -from-module=SOURCE  Copy the contents of the given module into the target
                                     directory before initialization.
              
                -get=true            Download any modules for this configuration.
              
                -input=true          Ask for input if necessary. If false, will error if
                                     input was required.
              
                -no-color            If specified, output won't contain any color.
              
                -plugin-dir          Directory containing plugin binaries. This overrides all
                                     default search paths for plugins, and prevents the
                                     automatic installation of plugins. This flag can be used
                                     multiple times.
              
                -reconfigure         Reconfigure the backend, ignoring any saved
                                     configuration.
              
                -upgrade=false       If installing modules (-get) or plugins, ignore
                                     previously-downloaded objects and install the
                                     latest version allowed within configured constraints.
      --- FAIL: TestAccAWSAcmCertificateDataSource_multipleIssued (2.43s)
```

### Steps to Reproduce

1. `TF_ACC=1 TF_ACC_TERRAFORM_PATH=/path/to/terraform0.15.0-alpha20210127 go test -v ./aws`

### References

- Initial Terraform CLI breaking change release: https://github.com/hashicorp/terraform/releases/tag/v0.15.0-alpha20210127
- Additional Terraform CLI breaking changes: https://github.com/hashicorp/terraform/pull/27664
- terraform-exec fixes (v.0.13.0): https://github.com/hashicorp/terraform-exec/pull/120
- Potential terraform-plugin-sdk fixes: https://github.com/hashicorp/terraform-plugin-sdk/pull/694

--- a/go.mod
+++ b/go.mod
@@ -25,7 +25,7 @@ require (
 	github.com/hashicorp/go-version v1.2.1
 	github.com/hashicorp/hcl/v2 v2.3.0
 	github.com/hashicorp/logutils v1.0.0
-	github.com/hashicorp/terraform-exec v0.12.0
+	github.com/hashicorp/terraform-exec v0.13.0
 	github.com/hashicorp/terraform-json v0.8.0
 	github.com/hashicorp/terraform-plugin-go v0.2.1
 	github.com/hashicorp/yamux v0.0.0-20181012175058-2f1d1f20f75d // indirect

--- a/internal/plugintest/helper.go
+++ b/internal/plugintest/helper.go
@@ -107,12 +107,6 @@ func (h *Helper) NewWorkingDir() (*WorkingDir, error) {
 		return nil, err
 	}
 
-	// symlink the provider source files into the base directory
-	err = symlinkDirectoriesOnly(h.sourceDir, dir)
-	if err != nil {
-		return nil, err
-	}
-
 	return &WorkingDir{
 		h:             h,
 		baseDir:       dir,

--- a/internal/plugintest/helper.go
+++ b/internal/plugintest/helper.go
@@ -4,6 +4,8 @@ import (
 	"fmt"
 	"io/ioutil"
 	"os"
+
+	"github.com/hashicorp/terraform-exec/tfexec"
 )
 
 const subprocessCurrentSigil = "4acd63807899403ca4859f5bb948d2c6"
@@ -107,8 +109,21 @@ func (h *Helper) NewWorkingDir() (*WorkingDir, error) {
 		return nil, err
 	}
 
+	// symlink the provider source files into the config directory
+	// e.g. testdata
+	err = symlinkDirectoriesOnly(h.sourceDir, dir)
+	if err != nil {
+		return nil, err
+	}
+
+	tf, err := tfexec.NewTerraform(dir, h.terraformExec)
+	if err != nil {
+		return nil, err
+	}
+
 	return &WorkingDir{
 		h:             h,
+		tf:            tf,
 		baseDir:       dir,
 		terraformExec: h.terraformExec,
 	}, nil

--- a/internal/plugintest/working_dir.go
+++ b/internal/plugintest/working_dir.go
@@ -102,7 +102,14 @@ func (wd *WorkingDir) SetConfig(cfg string) error {
 		return err
 	}
 
-	tf, err := tfexec.NewTerraform(wd.baseDir, wd.terraformExec)
+	// symlink the provider source files into the config directory
+	// e.g. testdata
+	err = symlinkDirectoriesOnly(wd.h.sourceDir, configDir)
+	if err != nil {
+		return err
+	}
+
+	tf, err := tfexec.NewTerraform(configDir, wd.terraformExec)
 	if err != nil {
 		return err
 	}
@@ -137,7 +144,7 @@ func (wd *WorkingDir) SetConfig(cfg string) error {
 // Any remote objects tracked by the state are not destroyed first, so this
 // will leave them dangling in the remote system.
 func (wd *WorkingDir) ClearState() error {
-	err := os.Remove(filepath.Join(wd.baseDir, "terraform.tfstate"))
+	err := os.Remove(filepath.Join(wd.configDir, "terraform.tfstate"))
 	if os.IsNotExist(err) {
 		return nil
 	}
@@ -160,24 +167,24 @@ func (wd *WorkingDir) Init() error {
 		return fmt.Errorf("must call SetConfig before Init")
 	}
 
-	return wd.tf.Init(context.Background(), tfexec.Reattach(wd.reattachInfo), tfexec.Dir(wd.configDir))
+	return wd.tf.Init(context.Background(), tfexec.Reattach(wd.reattachInfo))
 }
 
 func (wd *WorkingDir) planFilename() string {
-	return filepath.Join(wd.baseDir, "tfplan")
+	return filepath.Join(wd.configDir, "tfplan")
 }
 
 // CreatePlan runs "terraform plan" to create a saved plan file, which if successful
 // will then be used for the next call to Apply.
 func (wd *WorkingDir) CreatePlan() error {
-	_, err := wd.tf.Plan(context.Background(), tfexec.Reattach(wd.reattachInfo), tfexec.Refresh(false), tfexec.Out("tfplan"), tfexec.Dir(wd.configDir))
+	_, err := wd.tf.Plan(context.Background(), tfexec.Reattach(wd.reattachInfo), tfexec.Refresh(false), tfexec.Out("tfplan"))
 	return err
 }
 
 // CreateDestroyPlan runs "terraform plan -destroy" to create a saved plan
 // file, which if successful will then be used for the next call to Apply.
 func (wd *WorkingDir) CreateDestroyPlan() error {
-	_, err := wd.tf.Plan(context.Background(), tfexec.Reattach(wd.reattachInfo), tfexec.Refresh(false), tfexec.Out("tfplan"), tfexec.Destroy(true), tfexec.Dir(wd.configDir))
+	_, err := wd.tf.Plan(context.Background(), tfexec.Reattach(wd.reattachInfo), tfexec.Refresh(false), tfexec.Out("tfplan"), tfexec.Destroy(true))
 	return err
 }
 
@@ -189,17 +196,8 @@ func (wd *WorkingDir) Apply() error {
 	args := []tfexec.ApplyOption{tfexec.Reattach(wd.reattachInfo), tfexec.Refresh(false)}
 	if wd.HasSavedPlan() {
 		args = append(args, tfexec.DirOrPlan("tfplan"))
-	} else {
-		// we need to use a relative config dir here or we get an
-		// error about Terraform not having any configuration. See
-		// https://github.com/hashicorp/terraform-plugin-sdk/issues/495
-		// for more info.
-		configDir, err := wd.relativeConfigDir()
-		if err != nil {
-			return err
-		}
-		args = append(args, tfexec.DirOrPlan(configDir))
 	}
+
 	return wd.tf.Apply(context.Background(), args...)
 }
 
@@ -209,7 +207,7 @@ func (wd *WorkingDir) Apply() error {
 // If destroy fails then remote objects might still exist, and continue to
 // exist after a particular test is concluded.
 func (wd *WorkingDir) Destroy() error {
-	return wd.tf.Destroy(context.Background(), tfexec.Reattach(wd.reattachInfo), tfexec.Refresh(false), tfexec.Dir(wd.configDir))
+	return wd.tf.Destroy(context.Background(), tfexec.Reattach(wd.reattachInfo), tfexec.Refresh(false))
 }
 
 // HasSavedPlan returns true if there is a saved plan in the working directory. If
@@ -267,7 +265,7 @@ func (wd *WorkingDir) Import(resource, id string) error {
 
 // Refresh runs terraform refresh
 func (wd *WorkingDir) Refresh() error {
-	return wd.tf.Refresh(context.Background(), tfexec.Reattach(wd.reattachInfo), tfexec.State(filepath.Join(wd.baseDir, "terraform.tfstate")), tfexec.Dir(wd.configDir))
+	return wd.tf.Refresh(context.Background(), tfexec.Reattach(wd.reattachInfo), tfexec.State(filepath.Join(wd.configDir, "terraform.tfstate")))
 }
 
 // Schemas returns an object describing the provider schemas.

--- a/internal/plugintest/working_dir.go
+++ b/internal/plugintest/working_dir.go
@@ -13,6 +13,11 @@ import (
 	tfjson "github.com/hashicorp/terraform-json"
 )
 
+const (
+	ConfigFileName = "terraform_plugin_test.tf"
+	PlanFileName   = "tfplan"
+)
+
 // WorkingDir represents a distinct working directory that can be used for
 // running tests. Each test should construct its own WorkingDir by calling
 // NewWorkingDir or RequireNewWorkingDir on its package's singleton
@@ -25,9 +30,6 @@ type WorkingDir struct {
 
 	// baseArgs is arguments that should be appended to all commands
 	baseArgs []string
-
-	// configDir contains the singular config file generated for each test
-	configDir string
 
 	// tf is the instance of tfexec.Terraform used for running Terraform commands
 	tf *tfexec.Terraform
@@ -75,61 +77,31 @@ func (wd *WorkingDir) GetHelper() *Helper {
 	return wd.h
 }
 
-func (wd *WorkingDir) relativeConfigDir() (string, error) {
-	relPath, err := filepath.Rel(wd.baseDir, wd.configDir)
-	if err != nil {
-		return "", fmt.Errorf("Error determining relative path of configuration directory: %w", err)
-	}
-	return relPath, nil
-}
-
 // SetConfig sets a new configuration for the working directory.
 //
 // This must be called at least once before any call to Init, Plan, Apply, or
 // Destroy to establish the configuration. Any previously-set configuration is
 // discarded and any saved plan is cleared.
 func (wd *WorkingDir) SetConfig(cfg string) error {
-	// Each call to SetConfig creates a new directory under our baseDir.
-	// We create them within so that our final cleanup step will delete them
-	// automatically without any additional tracking.
-	configDir, err := ioutil.TempDir(wd.baseDir, "config")
-	if err != nil {
-		return err
-	}
-	configFilename := filepath.Join(configDir, "terraform_plugin_test.tf")
-	err = ioutil.WriteFile(configFilename, []byte(cfg), 0700)
-	if err != nil {
-		return err
-	}
-
-	// symlink the provider source files into the config directory
-	// e.g. testdata
-	err = symlinkDirectoriesOnly(wd.h.sourceDir, configDir)
-	if err != nil {
-		return err
-	}
-
-	tf, err := tfexec.NewTerraform(configDir, wd.terraformExec)
+	configFilename := filepath.Join(wd.baseDir, ConfigFileName)
+	err := ioutil.WriteFile(configFilename, []byte(cfg), 0700)
 	if err != nil {
 		return err
 	}
 
 	var mismatch *tfexec.ErrVersionMismatch
-	err = tf.SetDisablePluginTLS(true)
+	err = wd.tf.SetDisablePluginTLS(true)
 	if err != nil && !errors.As(err, &mismatch) {
 		return err
 	}
-	err = tf.SetSkipProviderVerify(true)
+	err = wd.tf.SetSkipProviderVerify(true)
 	if err != nil && !errors.As(err, &mismatch) {
 		return err
 	}
 
 	if p := os.Getenv("TF_ACC_LOG_PATH"); p != "" {
-		tf.SetLogPath(p)
+		wd.tf.SetLogPath(p)
 	}
-
-	wd.configDir = configDir
-	wd.tf = tf
 
 	// Changing configuration invalidates any saved plan.
 	err = wd.ClearPlan()
@@ -144,7 +116,7 @@ func (wd *WorkingDir) SetConfig(cfg string) error {
 // Any remote objects tracked by the state are not destroyed first, so this
 // will leave them dangling in the remote system.
 func (wd *WorkingDir) ClearState() error {
-	err := os.Remove(filepath.Join(wd.configDir, "terraform.tfstate"))
+	err := os.Remove(filepath.Join(wd.baseDir, "terraform.tfstate"))
 	if os.IsNotExist(err) {
 		return nil
 	}
@@ -163,28 +135,32 @@ func (wd *WorkingDir) ClearPlan() error {
 // Init runs "terraform init" for the given working directory, forcing Terraform
 // to use the current version of the plugin under test.
 func (wd *WorkingDir) Init() error {
-	if wd.configDir == "" {
+	if _, err := os.Stat(wd.configFilename()); err != nil {
 		return fmt.Errorf("must call SetConfig before Init")
 	}
 
 	return wd.tf.Init(context.Background(), tfexec.Reattach(wd.reattachInfo))
 }
 
+func (wd *WorkingDir) configFilename() string {
+	return filepath.Join(wd.baseDir, ConfigFileName)
+}
+
 func (wd *WorkingDir) planFilename() string {
-	return filepath.Join(wd.configDir, "tfplan")
+	return filepath.Join(wd.baseDir, PlanFileName)
 }
 
 // CreatePlan runs "terraform plan" to create a saved plan file, which if successful
 // will then be used for the next call to Apply.
 func (wd *WorkingDir) CreatePlan() error {
-	_, err := wd.tf.Plan(context.Background(), tfexec.Reattach(wd.reattachInfo), tfexec.Refresh(false), tfexec.Out("tfplan"))
+	_, err := wd.tf.Plan(context.Background(), tfexec.Reattach(wd.reattachInfo), tfexec.Refresh(false), tfexec.Out(PlanFileName))
 	return err
 }
 
 // CreateDestroyPlan runs "terraform plan -destroy" to create a saved plan
 // file, which if successful will then be used for the next call to Apply.
 func (wd *WorkingDir) CreateDestroyPlan() error {
-	_, err := wd.tf.Plan(context.Background(), tfexec.Reattach(wd.reattachInfo), tfexec.Refresh(false), tfexec.Out("tfplan"), tfexec.Destroy(true))
+	_, err := wd.tf.Plan(context.Background(), tfexec.Reattach(wd.reattachInfo), tfexec.Refresh(false), tfexec.Out(PlanFileName), tfexec.Destroy(true))
 	return err
 }
 
@@ -195,7 +171,7 @@ func (wd *WorkingDir) CreateDestroyPlan() error {
 func (wd *WorkingDir) Apply() error {
 	args := []tfexec.ApplyOption{tfexec.Reattach(wd.reattachInfo), tfexec.Refresh(false)}
 	if wd.HasSavedPlan() {
-		args = append(args, tfexec.DirOrPlan("tfplan"))
+		args = append(args, tfexec.DirOrPlan(PlanFileName))
 	}
 
 	return wd.tf.Apply(context.Background(), args...)
@@ -260,12 +236,12 @@ func (wd *WorkingDir) State() (*tfjson.State, error) {
 
 // Import runs terraform import
 func (wd *WorkingDir) Import(resource, id string) error {
-	return wd.tf.Import(context.Background(), resource, id, tfexec.Config(wd.configDir), tfexec.Reattach(wd.reattachInfo))
+	return wd.tf.Import(context.Background(), resource, id, tfexec.Config(wd.baseDir), tfexec.Reattach(wd.reattachInfo))
 }
 
 // Refresh runs terraform refresh
 func (wd *WorkingDir) Refresh() error {
-	return wd.tf.Refresh(context.Background(), tfexec.Reattach(wd.reattachInfo), tfexec.State(filepath.Join(wd.configDir, "terraform.tfstate")))
+	return wd.tf.Refresh(context.Background(), tfexec.Reattach(wd.reattachInfo), tfexec.State(filepath.Join(wd.baseDir, "terraform.tfstate")))
 }
 
 // Schemas returns an object describing the provider schemas.

